mqtt/mqtt-broker/src/persist.rs

```diff
@@ -135,6 +138,209 @@ impl FileFormat for BincodeFormat {
     }
 }
 
+use crate::session::SessionState;
+use crate::subscription::Subscription;
+use crate::ClientId;
+use bytes::Bytes;
+use mqtt3::proto::Publication;
+use serde::ser::SerializeMap;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::hash_map::DefaultHasher;
+use std::collections::{HashMap, VecDeque};
+use std::hash::{Hash, Hasher};
+use std::iter::FromIterator;
+
+pub struct ConsolidatedStateFormat;
+
+impl ConsolidatedStateFormat {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+impl FileFormat for ConsolidatedStateFormat {
+    type Error = Error;
+
+    fn load<R: Read>(&self, reader: R) -> Result<BrokerState, Self::Error> {
+        let decoder = GzDecoder::new(reader);
+        fail_point!("bincodeformat.load.deserialize_from", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        let state = bincode::deserialize_from(decoder)
+            .context(ErrorKind::Persist(ErrorReason::Deserialize))?;
+
+        let state = Consolidator::resolve_state(state);
+        Ok(state)
+    }
+
+    fn store<W: Write>(&self, writer: W, state: BrokerState) -> Result<(), Self::Error> {
+        let state = Consolidator::consolidate_state(state);
+
+        let encoder = GzEncoder::new(writer, Compression::default());
+        fail_point!("bincodeformat.store.serialize_into", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        bincode::serialize_into(encoder, &state)
+            .context(ErrorKind::Persist(ErrorReason::Serialize))?;
+        Ok(())
+    }
+}
+
+struct Consolidator {
```

Comment:
I believe you don't even need this struct. You can improve ergonomics of the code doing:
1. Implement `From<T> for U` trait for all parts of the consolidation state (state, sessions, publications)
2. Use `HashMap<Uuid, Publication>` from the ConsolidatedState to collect all payloads
3. Try to use as much as you can existing public API to construct state items
4. Because all ConsolidatedXXXX structs are persistor implementation details only, I suggest to rename them with the same Name as original structs.

Reply:
1. I can't use from, since the state (hashmap) has to be passed in
1. Changed to let hashmap handle de-duping and hashing
1. Existing api? 
1. How would that work? `consolidate_state(state: mqtt::BrokerState) -> BrokerState` seems more confusing

Reply:
Please check code from my comment below